docs/modules/ROOT/pages/deploy_related/integration_keystone.adoc

```diff
@@ -0,0 +1,53 @@
+= Integration with OpenStack Keystone
+v0.1, 2020-05-28
+:toc: right
+
+NOTE: This document is under development
+
+== Overview
+
+Alcor API currently requires a project id in the API Url.Neutron 2.0 APIs, however, don't need to provide a project id as
+it generates the project id based on the Authentication token in the X-Auth-Token request header, when Keystone is enabled.
+
+Reference: https://docs.openstack.org/api-ref/network/v2/
+"The Networking API v2.0 uses the OpenStack Identity service as the default authentication service.
+When Keystone is enabled, users that submit requests to the OpenStack Networking service must provide an authentication
+token in X-Auth-Token request header. You obtain the token by authenticating to the Keystone endpoint.
+
+When Keystone is enabled, the project_id attribute is not required in create requests
+because the project ID is derived from the authentication token."
+
+[#system-requirements]
+== Integration Requirements
+
+. Identify Neutron's algorithm that generates project id based on X-Auth-Token.
+. Make changes to API Gateway to support Authentication & Authorization with KeyStone, and support same way how project id is generated based on X-Auth-Token in API Gateway.
+. Integration testing with KeyStone.
+
+== Review of OpenStack Workflow
+https://docs.openstack.org/api-ref/identity/v3/index.html#check-token[Openstack keystone apis]
+
+In openstack neutron project, it use keystonemiddle to verify request token. .
+Verify request token have few steps.
+
+. get available keystone endpoints from keystone server
+. request local token from keystone server with keystone client conf in neutron.conf
+. verify request token from keystone server
+
+== Required Changes
+
+. Add a KeystoneAuthWebFilter class to fetch all api-gateway request token and verify token from keystone server and get project_id from verify api response.
+. Add keystone conf in application.properties file (https://docs.openstack.org/keystone/latest/user/supported_clients.html[keystone config example])
+
+== KeystoneAuthWebFilter Workflow
+image::keystone_filter_workflow.jpg["KeystoneAuthWebFilter Workflow", width=1024]
```

Comment:
I have a few comments: 
* It seems that Steps 2 - 5 are not required for every customer request, meaning that every API Gateway instance could proactively retrieve the "alcor token" when the instance launches. Is this understanding right? If so, we could split the diagram into two diagrams, one for the preparation phase of Alcor and the other for customer request. 
* Could you elaborate what "Alcor token" is, what will it be used for?
* We could need a cache layer to store the Alcor token and customer tokens as well. Could you add the cache into this diagram and details in the design doc? Thank you.
* Some minor typos in the diagram:
1. Gateway Api Layter => Gateway API Layer
2. Micro Service 1 => Microservice 1
3. Request Realy Service => do you mean "Request Real Service" or "Request Ready Service"

Reply:
- An Alcor token is used for verifying customer request token, it is different from request token that has more permissions than a request token. An Alcor is generate from keystone by using Alcor local keystone properties.
- Steps 2-3 are required for the first customer request come, it will init keystone available endpoint. But steps 4-5 are required for every customer when the Alcor token is null or expired.

Reply:
> * An Alcor token is used for verifying customer request token, it is different from request token that has more permissions than a request token. An Alcor is generate from keystone by using Alcor local keystone properties.

I see. Alcor token is a token owned by infra and used to verify the request token.

> * Steps 2-3 are required for the first customer request come, it will init keystone available endpoint. But steps 4-5 are required for every customer when the Alcor token is null or expired.

Okay. Neither Steps 2-3 nor Steps 4-5 are executed for every customer request. Instead, they will be executed proactively. For example, 
* Steps 2-3 could be triggered when we launch a new Alcor instance, no need to wait until the first call comes.
* Steps 4-5 could be proactively triggered when the existing Alcor token is null or close to expiry, meaning that we need to maintain at least two tokens at that time. We can't wait until the "first" customer request arrives, as there might be multiple such calls arriving almost at the same time.

Reply:
Ok,
- Alcor could execute steps 2-5 to get available endpoints and token when it starts.
- I think the remaining two tokens are the same as the remaining one token. Because a token have a solid expire time, two tokens will expire at the same time. Alcor could request a new token when the current token expiring soon.

Reply:
Saw the new diagrams. Awesome!